File: examples/lua/comprehensive/01_basic_plugin_management.lua
#!/usr/bin/env lua
--[[
QtForge Lua Bindings Example: Basic Plugin Management

This example demonstrates the fundamental plugin management operations using
the QtForge Lua bindings. It covers:

1. Creating and configuring a PluginManager
2. Loading and unloading plugins
3. Querying plugin information
4. Managing plugin lifecycle
5. Error handling and best practices

Prerequisites:
- QtForge built with Lua bindings enabled
- Lua 5.1 or later (compatible with LuaJIT)
- sol2 library for C++ binding
]]

-- Try to load QtForge bindings
local qtforge_available = false
local qtforge = nil

local success, result = pcall(require, "qtforge")
if success then
    qtforge = result
    qtforge_available = true
    print("✅ QtForge Lua bindings loaded successfully")
    if qtforge.version then
        print(string.format("QtForge version: %s", qtforge.version))
    end
else
    print(string.format("❌ Failed to load QtForge: %s", result))
    print("Make sure QtForge is built with Lua bindings enabled")
    os.exit(1)
end

-- PluginManagerExample class
local PluginManagerExample = {}
PluginManagerExample.__index = PluginManagerExample

function PluginManagerExample:new()
    local obj = {
        manager = nil,
        registry = nil,
        loader = nil
    }
    setmetatable(obj, PluginManagerExample)
    obj:setup_components()
    return obj
end

function PluginManagerExample:setup_components()
    print("\n🔧 Setting up plugin management components...")
    
    -- Create PluginManager
    if qtforge.core and qtforge.core.create_plugin_manager then
        self.manager = qtforge.core.create_plugin_manager()
        print("✅ PluginManager created via factory function")
    elseif qtforge.core and qtforge.core.PluginManager then
        self.manager = qtforge.core.PluginManager()
        print("✅ PluginManager created")
    else
        error("PluginManager not available")
    end
    
    -- Create PluginRegistry
    if qtforge.core and qtforge.core.create_plugin_registry then
        self.registry = qtforge.core.create_plugin_registry()
        print("✅ PluginRegistry created via factory function")
    elseif qtforge.core and qtforge.core.PluginRegistry then
        self.registry = qtforge.core.PluginRegistry()
        print("✅ PluginRegistry created")
    end
    
    -- Create PluginLoader
    if qtforge.core and qtforge.core.create_plugin_loader then
        self.loader = qtforge.core.create_plugin_loader()
        print("✅ PluginLoader created via factory function")
    elseif qtforge.core and qtforge.core.PluginLoader then
        self.loader = qtforge.core.PluginLoader()
        print("✅ PluginLoader created")
    end
end

function PluginManagerExample:demonstrate_basic_operations()
    print("\n📋 Demonstrating basic plugin operations...")
    
    -- Check initial state
    if self.manager and self.manager.get_plugin_count then
        local initial_count = self.manager:get_plugin_count()
        print(string.format("Initial plugin count: %d", initial_count))
    end
    
    -- Get all plugins (should be empty initially)
    if self.manager and self.manager.get_all_plugins then
        local plugins = self.manager:get_all_plugins()
        print(string.format("Initial plugins list length: %d", #plugins))
    end
    
    -- Test plugin existence check
    if self.manager and self.manager.has_plugin then
        local has_test_plugin = self.manager:has_plugin("test_plugin")
        print(string.format("Has 'test_plugin': %s", tostring(has_test_plugin)))
    end
    
    -- Test getting non-existent plugin
    if self.manager and self.manager.get_plugin then
        local plugin = self.manager:get_plugin("non_existent_plugin")
        print(string.format("Get non-existent plugin result: %s", tostring(plugin)))
    end
end

function PluginManagerExample:demonstrate_plugin_loading()
    print("\n🔌 Demonstrating plugin loading...")
    
    -- Create some example plugin paths (these won't exist, but demonstrate the API)
    local example_plugins = {
        "/path/to/example_plugin.so",
        "/path/to/another_plugin.dll",
        "relative/path/plugin.dylib"
    }
    
    for _, plugin_path in ipairs(example_plugins) do
        print(string.format("\nAttempting to load plugin: %s", plugin_path))
        
        local success, result = pcall(function()
            if self.manager and self.manager.load_plugin then
                return self.manager:load_plugin(plugin_path)
            elseif self.loader and self.loader.load_plugin then
                return self.loader:load_plugin(plugin_path)
            else
                return "Plugin loading not available"
            end
        end)
        
        if success then
            print(string.format("Load result: %s", tostring(result)))
        else
            print(string.format("Expected error (file doesn't exist): %s", result))
        end
    end
end

function PluginManagerExample:demonstrate_plugin_queries()
    print("\n🔍 Demonstrating plugin queries...")
    
    -- Query plugins by capability
    if self.manager and self.manager.get_plugins_by_capability then
        local capabilities = {}
        
        -- Collect available capabilities
        if qtforge.core and qtforge.core.PluginCapability then
            local cap_names = {"Service", "UI", "Network", "DataProcessor"}
            for _, name in ipairs(cap_names) do
                if qtforge.core.PluginCapability[name] then
                    table.insert(capabilities, qtforge.core.PluginCapability[name])
                end
            end
        end
        
        for _, capability in ipairs(capabilities) do
            local success, plugins = pcall(function()
                return self.manager:get_plugins_by_capability(capability)
            end)
            
            if success then
                print(string.format("Plugins with capability %s: %d", tostring(capability), #plugins))
            else
                print(string.format("Plugin capability query error: %s", plugins))
            end
        end
    end
    
    -- Test registry operations if available
    if self.registry then
        print("\n📚 Testing registry operations...")
        
        if self.registry.size then
            local size = self.registry:size()
            print(string.format("Registry size: %d", size))
        end
        
        if self.registry.get_all_plugins then
            local all_plugins = self.registry:get_all_plugins()
            print(string.format("Registry plugins: %d", #all_plugins))
        end
    end
end

function PluginManagerExample:demonstrate_plugin_states()
    print("\n🔄 Demonstrating plugin states...")
    
    -- Show available plugin states
    if qtforge.core and qtforge.core.PluginState then
        print("Available plugin states:")
        local state_names = {"Unloaded", "Loaded", "Initialized", "Running", "Stopped", "Error"}
        for _, name in ipairs(state_names) do
            if qtforge.core.PluginState[name] then
                local state = qtforge.core.PluginState[name]
                print(string.format("  - %s: %s", name, tostring(state)))
            end
        end
    end
    
    -- Show available plugin capabilities
    if qtforge.core and qtforge.core.PluginCapability then
        print("\nAvailable plugin capabilities:")
        local cap_names = {"None", "Service", "UI", "Network", "DataProcessor", "Scripting"}
        for _, name in ipairs(cap_names) do
            if qtforge.core.PluginCapability[name] then
                local capability = qtforge.core.PluginCapability[name]
                print(string.format("  - %s: %s", name, tostring(capability)))
            end
        end
    end
    
    -- Show available plugin priorities
    if qtforge.core and qtforge.core.PluginPriority then
        print("\nAvailable plugin priorities:")
        local priority_names = {"Lowest", "Low", "Normal", "High", "Highest"}
        for _, name in ipairs(priority_names) do
            if qtforge.core.PluginPriority[name] then
                local priority = qtforge.core.PluginPriority[name]
                print(string.format("  - %s: %s", name, tostring(priority)))
            end
        end
    end
end

function PluginManagerExample:demonstrate_dependency_management()
    print("\n🔗 Demonstrating dependency management...")
    
    -- Create dependency resolver if available
    local resolver = nil
    local success, result = pcall(function()
        if qtforge.core and qtforge.core.create_plugin_dependency_resolver then
            return qtforge.core.create_plugin_dependency_resolver()
        elseif qtforge.core and qtforge.core.PluginDependencyResolver then
            return qtforge.core.PluginDependencyResolver()
        else
            error("PluginDependencyResolver not available")
        end
    end)
    
    if success then
        resolver = result
        print("✅ PluginDependencyResolver created")
    else
        print(string.format("⚠️  Dependency resolver not available: %s", result))
        return
    end
    
    if resolver then
        -- Test dependency operations
        local operations = {
            {"get_load_order", "Current load order"},
            {"has_circular_dependencies", "Has circular dependencies"},
            {"get_dependency_graph", "Dependency graph"}
        }
        
        for _, op in ipairs(operations) do
            local method_name, description = op[1], op[2]
            if resolver[method_name] then
                local success, result = pcall(function()
                    return resolver[method_name](resolver)
                end)
                
                if success then
                    print(string.format("%s: %s", description, tostring(result)))
                else
                    print(string.format("%s error: %s", description, result))
                end
            end
        end
    end
end

function PluginManagerExample:demonstrate_error_handling()
    print("\n⚠️  Demonstrating error handling...")
    
    -- Test various error conditions
    local error_tests = {
        {
            name = "Loading non-existent plugin",
            func = function()
                if self.manager and self.manager.load_plugin then
                    return self.manager:load_plugin("/non/existent/path.so")
                end
                return nil
            end
        },
        {
            name = "Getting non-existent plugin",
            func = function()
                if self.manager and self.manager.get_plugin then
                    return self.manager:get_plugin("non_existent")
                end
                return nil
            end
        },
        {
            name = "Unloading non-existent plugin",
            func = function()
                if self.manager and self.manager.unload_plugin then
                    return self.manager:unload_plugin("non_existent")
                end
                return nil
            end
        }
    }
    
    for _, test in ipairs(error_tests) do
        print(string.format("\nTesting: %s", test.name))
        local success, result = pcall(test.func)
        
        if success then
            print(string.format("Result: %s (type: %s)", tostring(result), type(result)))
        else
            print(string.format("Exception caught: %s", result))
        end
    end
end

function PluginManagerExample:cleanup()
    print("\n🧹 Cleaning up resources...")
    
    -- Clear manager if available
    if self.manager and self.manager.clear then
        local success, result = pcall(function()
            self.manager:clear()
        end)
        
        if success then
            print("✅ PluginManager cleared")
        else
            print(string.format("Manager cleanup error: %s", result))
        end
    end
    
    -- Clear registry if available
    if self.registry and self.registry.clear then
        local success, result = pcall(function()
            self.registry:clear()
        end)
        
        if success then
            print("✅ PluginRegistry cleared")
        else
            print(string.format("Registry cleanup error: %s", result))
        end
    end
end

function PluginManagerExample:run_complete_example()
    print("🚀 QtForge Lua Bindings - Basic Plugin Management Example")
    print(string.rep("=", 70))
    
    local success, error_msg = pcall(function()
        self:demonstrate_basic_operations()
        self:demonstrate_plugin_loading()
        self:demonstrate_plugin_queries()
        self:demonstrate_plugin_states()
        self:demonstrate_dependency_management()
        self:demonstrate_error_handling()
    end)
    
    -- Always cleanup
    self:cleanup()
    
    if success then
        print("\n✅ Basic plugin management example completed successfully!")
        return 0
    else
        print(string.format("\n❌ Example failed with error: %s", error_msg))
        return 1
    end
end

-- Main function
local function main()
    local success, result = pcall(function()
        local example = PluginManagerExample:new()
        return example:run_complete_example()
    end)
    
    if success then
        return result
    else
        print(string.format("❌ Failed to run example: %s", result))
        return 1
    end
end

-- Run the example
local exit_code = main()
os.exit(exit_code)

File: examples/python/comprehensive/01_basic_plugin_management.py
#!/usr/bin/env python3
"""
QtForge Python Bindings Example: Basic Plugin Management

This example demonstrates the fundamental plugin management operations using
the QtForge Python bindings. It covers:

1. Creating and configuring a PluginManager
2. Loading and unloading plugins
3. Querying plugin information
4. Managing plugin lifecycle
5. Error handling and best practices

Prerequisites:
- QtForge built with Python bindings enabled
- Python 3.8 or later
"""
# type: ignore

import sys
import os
from pathlib import Path
from typing import Any, Optional

# Add the build directory to Python path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../../build'))

try:
    import qtforge
    import qtforge.core as core
    print("✅ QtForge Python bindings loaded successfully")
    print(f"QtForge version: {qtforge.__version__ if hasattr(qtforge, '__version__') else 'Unknown'}")
except ImportError as e:
    print(f"❌ Failed to import QtForge: {e}")
    print("Make sure QtForge is built with Python bindings enabled")
    sys.exit(1)


class PluginManagerExample:
    """Comprehensive example of plugin management operations."""
    
    def __init__(self) -> None:
        """Initialize the plugin manager example."""
        self.manager: Optional[Any] = None
        self.registry: Optional[Any] = None
        self.loader: Optional[Any] = None
        self.setup_components()
    
    def setup_components(self) -> None:
        """Set up the core plugin management components."""
        print("\n🔧 Setting up plugin management components...")
        
        try:
            # Create PluginManager
            if hasattr(core, 'PluginManager'):
                self.manager = core.PluginManager()
                print("✅ PluginManager created")
            elif hasattr(core, 'create_plugin_manager'):
                self.manager = core.create_plugin_manager()
                print("✅ PluginManager created via factory function")
            else:
                raise RuntimeError("PluginManager not available")
            
            # Create PluginRegistry
            if hasattr(core, 'PluginRegistry'):
                self.registry = core.PluginRegistry()
                print("✅ PluginRegistry created")
            elif hasattr(core, 'create_plugin_registry'):
                self.registry = core.create_plugin_registry()
                print("✅ PluginRegistry created via factory function")
            
            # Create PluginLoader
            if hasattr(core, 'PluginLoader'):
                self.loader = core.PluginLoader()
                print("✅ PluginLoader created")
            elif hasattr(core, 'create_plugin_loader'):
                self.loader = core.create_plugin_loader()
                print("✅ PluginLoader created via factory function")
            
        except Exception as e:
            print(f"❌ Failed to setup components: {e}")
            raise
    
    def demonstrate_basic_operations(self) -> None:
        """Demonstrate basic plugin management operations."""
        print("\n📋 Demonstrating basic plugin operations...")
        
        # Check initial state
        if self.manager and hasattr(self.manager, 'get_plugin_count'):
            initial_count = self.manager.get_plugin_count()
            print(f"Initial plugin count: {initial_count}")
        
        # Get all plugins (should be empty initially)
        if self.manager and hasattr(self.manager, 'get_all_plugins'):
            plugins = self.manager.get_all_plugins()
            print(f"Initial plugins list length: {len(plugins)}")
        
        # Test plugin existence check
        if self.manager and hasattr(self.manager, 'has_plugin'):
            has_test_plugin = self.manager.has_plugin("test_plugin")
            print(f"Has 'test_plugin': {has_test_plugin}")
        
        # Test getting non-existent plugin
        if self.manager and hasattr(self.manager, 'get_plugin'):
            plugin = self.manager.get_plugin("non_existent_plugin")
            print(f"Get non-existent plugin result: {plugin}")
    
    def demonstrate_plugin_loading(self) -> None:
        """Demonstrate plugin loading operations."""
        print("\n🔌 Demonstrating plugin loading...")
        
        # Create some example plugin paths (these won't exist, but demonstrate the API)
        example_plugins = [
            "/path/to/example_plugin.so",
            "/path/to/another_plugin.dll",
            "relative/path/plugin.dylib"
        ]
        
        for plugin_path in example_plugins:
            print(f"\nAttempting to load plugin: {plugin_path}")
            
            try:
                if self.manager and hasattr(self.manager, 'load_plugin'):
                    result = self.manager.load_plugin(plugin_path)
                    print(f"Load result: {result}")
                elif self.loader and hasattr(self.loader, 'load_plugin'):
                    result = self.loader.load_plugin(plugin_path)
                    print(f"Load result: {result}")
                else:
                    print("⚠️  Plugin loading not available")
                    
            except Exception as e:
                print(f"Expected error (file doesn't exist): {e}")
    
    def demonstrate_plugin_queries(self) -> None:
        """Demonstrate plugin query operations."""
        print("\n🔍 Demonstrating plugin queries...")
        
        # Query plugins by capability
        if self.manager and hasattr(self.manager, 'get_plugins_by_capability'):
            try:
                # Test different capabilities if available
                capabilities = []
                if hasattr(core, 'PluginCapability'):
                    cap_attrs = ['Service', 'UI', 'Network', 'DataProcessor']
                    for attr in cap_attrs:
                        if hasattr(core.PluginCapability, attr):
                            capabilities.append(getattr(core.PluginCapability, attr))
                
                for capability in capabilities:
                    plugins = self.manager.get_plugins_by_capability(capability)
                    print(f"Plugins with capability {capability}: {len(plugins)}")
                    
            except Exception as e:
                print(f"Plugin capability query error: {e}")
        
        # Test registry operations if available
        if self.registry:
            print("\n📚 Testing registry operations...")
            
            if hasattr(self.registry, 'size'):
                size = self.registry.size()
                print(f"Registry size: {size}")
            
            if hasattr(self.registry, 'get_all_plugins'):
                all_plugins = self.registry.get_all_plugins()
                print(f"Registry plugins: {len(all_plugins)}")
    
    def demonstrate_plugin_states(self) -> None:
        """Demonstrate plugin state management."""
        print("\n🔄 Demonstrating plugin states...")
        
        # Show available plugin states
        if hasattr(core, 'PluginState'):
            print("Available plugin states:")
            state_attrs = ['Unloaded', 'Loaded', 'Initialized', 'Running', 'Stopped', 'Error']
            for attr in state_attrs:
                if hasattr(core.PluginState, attr):
                    state = getattr(core.PluginState, attr)
                    print(f"  - {attr}: {state}")
        
        # Show available plugin capabilities
        if hasattr(core, 'PluginCapability'):
            print("\nAvailable plugin capabilities:")
            cap_attrs = ['None', 'Service', 'UI', 'Network', 'DataProcessor', 'Scripting']
            for attr in cap_attrs:
                if hasattr(core.PluginCapability, attr):
                    capability = getattr(core.PluginCapability, attr)
                    print(f"  - {attr}: {capability}")
        
        # Show available plugin priorities
        if hasattr(core, 'PluginPriority'):
            print("\nAvailable plugin priorities:")
            priority_attrs = ['Lowest', 'Low', 'Normal', 'High', 'Highest']
            for attr in priority_attrs:
                if hasattr(core.PluginPriority, attr):
                    priority = getattr(core.PluginPriority, attr)
                    print(f"  - {attr}: {priority}")
    
    def demonstrate_dependency_management(self) -> None:
        """Demonstrate plugin dependency management."""
        print("\n🔗 Demonstrating dependency management...")
        
        # Create dependency resolver if available
        resolver = None
        try:
            if hasattr(core, 'PluginDependencyResolver'):
                resolver = core.PluginDependencyResolver()
                print("✅ PluginDependencyResolver created")
            elif hasattr(core, 'create_plugin_dependency_resolver'):
                resolver = core.create_plugin_dependency_resolver()
                print("✅ PluginDependencyResolver created via factory")
        except Exception as e:
            print(f"⚠️  Dependency resolver not available: {e}")
            return
        
        if resolver:
            # Test dependency operations
            try:
                if hasattr(resolver, 'get_load_order'):
                    load_order = resolver.get_load_order()
                    print(f"Current load order: {load_order}")
                
                if hasattr(resolver, 'has_circular_dependencies'):
                    has_circular = resolver.has_circular_dependencies()
                    print(f"Has circular dependencies: {has_circular}")
                
                if hasattr(resolver, 'get_dependency_graph'):
                    graph = resolver.get_dependency_graph()
                    print(f"Dependency graph type: {type(graph)}")
                
            except Exception as e:
                print(f"Dependency operation error: {e}")
    
    def demonstrate_lifecycle_management(self) -> None:
        """Demonstrate plugin lifecycle management."""
        print("\n♻️  Demonstrating lifecycle management...")
        
        # Test lifecycle configuration if available
        if hasattr(core, 'PluginLifecycleConfig'):
            try:
                config = core.PluginLifecycleConfig()
                print("✅ PluginLifecycleConfig created")
                
                # Configure lifecycle settings
                if hasattr(config, 'initialization_timeout'):
                    config.initialization_timeout = 5000  # type: ignore  # 5 seconds
                    print(f"Set initialization timeout: {config.initialization_timeout}ms")
                
                if hasattr(config, 'enable_health_monitoring'):
                    config.enable_health_monitoring = True
                    print(f"Health monitoring enabled: {config.enable_health_monitoring}")
                
                if hasattr(config, 'auto_restart_on_failure'):
                    config.auto_restart_on_failure = True
                    print(f"Auto-restart enabled: {config.auto_restart_on_failure}")
                
                # Test JSON serialization
                if hasattr(config, 'to_json'):
                    json_data = config.to_json()
                    print(f"Config serialized to JSON: {type(json_data)}")
                
            except Exception as e:
                print(f"Lifecycle configuration error: {e}")
        
        # Show lifecycle events if available
        if hasattr(core, 'PluginLifecycleEvent'):
            print("\nAvailable lifecycle events:")
            event_attrs = [
                'BeforeInitialize', 'AfterInitialize', 'BeforeShutdown', 'AfterShutdown',
                'StateChanged', 'Error', 'HealthCheck'
            ]
            for attr in event_attrs:
                if hasattr(core.PluginLifecycleEvent, attr):
                    event = getattr(core.PluginLifecycleEvent, attr)
                    print(f"  - {attr}: {event}")
    
    def demonstrate_error_handling(self) -> None:
        """Demonstrate proper error handling patterns."""
        print("\n⚠️  Demonstrating error handling...")
        
        # Test various error conditions
        error_tests = [
            ("Loading non-existent plugin", lambda: self.manager.load_plugin("/non/existent/path.so") if self.manager and hasattr(self.manager, 'load_plugin') else None),
            ("Getting non-existent plugin", lambda: self.manager.get_plugin("non_existent") if self.manager and hasattr(self.manager, 'get_plugin') else None),
            ("Unloading non-existent plugin", lambda: self.manager.unload_plugin("non_existent") if self.manager and hasattr(self.manager, 'unload_plugin') else None),
        ]
        
        for test_name, test_func in error_tests:
            print(f"\nTesting: {test_name}")
            try:
                result = test_func()
                print(f"Result: {result} (type: {type(result)})")
            except Exception as e:
                print(f"Exception caught: {type(e).__name__}: {e}")
    
    def cleanup(self) -> None:
        """Clean up resources."""
        print("\n🧹 Cleaning up resources...")
        
        # Clear manager if available
        if self.manager and hasattr(self.manager, 'clear'):
            try:
                self.manager.clear()
                print("✅ PluginManager cleared")
            except Exception as e:
                print(f"Manager cleanup error: {e}")
        
        # Clear registry if available
        if self.registry and hasattr(self.registry, 'clear'):
            try:
                self.registry.clear()
                print("✅ PluginRegistry cleared")
            except Exception as e:
                print(f"Registry cleanup error: {e}")
    
    def run_complete_example(self) -> int:
        """Run the complete plugin management example."""
        print("🚀 QtForge Python Bindings - Basic Plugin Management Example")
        print("=" * 70)
        
        try:
            self.demonstrate_basic_operations()
            self.demonstrate_plugin_loading()
            self.demonstrate_plugin_queries()
            self.demonstrate_plugin_states()
            self.demonstrate_dependency_management()
            self.demonstrate_lifecycle_management()
            self.demonstrate_error_handling()
            
        except Exception as e:
            print(f"\n❌ Example failed with error: {e}")
            return 1
        
        finally:
            self.cleanup()
        
        print("\n✅ Basic plugin management example completed successfully!")
        return 0


def main() -> int:
    """Main entry point for the example."""
    try:
        example = PluginManagerExample()
        return example.run_complete_example()
    except Exception as e:
        print(f"❌ Failed to run example: {e}")
        return 1


if __name__ == "__main__":
    sys.exit(main())
